STM32/Libraries/USBprint/Src/usb_cdc_fops.c

```diff
@@ -0,0 +1,275 @@
+    /*                                        4 - Space                            */
+    /* 6      | bDataBits  |   1   | Number Data bits (5, 6, 7, 8 or 16).          */
+    /*******************************************************************************/
+    case CDC_SET_LINE_CODING:
+        LineCoding.bitrate = (uint32_t)(pbuf[0] | (pbuf[1] << 8) |
+                (pbuf[2] << 16) | (pbuf[3] << 24));
+        LineCoding.format = pbuf[4];
+        LineCoding.paritytype = pbuf[5];
+        LineCoding.datatype = pbuf[6];
+    break;
+
+    case CDC_GET_LINE_CODING:
+        pbuf[0] = (uint8_t)(LineCoding.bitrate);
+        pbuf[1] = (uint8_t)(LineCoding.bitrate >> 8);
+        pbuf[2] = (uint8_t)(LineCoding.bitrate >> 16);
+        pbuf[3] = (uint8_t)(LineCoding.bitrate >> 24);
+        pbuf[4] = LineCoding.format;
+        pbuf[5] = LineCoding.paritytype;
+        pbuf[6] = LineCoding.datatype;
+    break;
+
+    case CDC_SET_CONTROL_LINE_STATE:
+        usb_cdc_if.isComPortOpen = (((USBD_SetupReqTypedef *) pbuf)->wValue & 0x0001) != 0;
+    break;
+
+    case CDC_SEND_BREAK:    break;
+    default:                break;
+  }
+
+  return (USBD_OK);
+}
+
+/**
+  * @brief  Data received over USB OUT endpoint are sent over CDC interface
+  *         through this function.
+  *
+  *         @note
+  *         This function will issue a NAK packet on any OUT packet received on
+  *         USB endpoint until exiting this function. If you exit this function
+  *         before transfer is complete on CDC interface (ie. using DMA controller)
+  *         it will result in receiving more data while previous ones are still
+  *         not sent.
+  *
+  * @param  Buf: Buffer of data to be received
+  * @param  Len: Number of data received (in bytes)
+  * @retval Result of the operation: USBD_OK if all operations are OK else USBD_FAIL
+  */
+static int8_t CDC_Receive_FS(uint8_t* Buf, uint32_t *Len)
+{
+    USBD_CDC_SetRxBuffer(&hUsbDeviceFS, &Buf[0]);
+    USBD_CDC_ReceivePacket(&hUsbDeviceFS);
+
+    uint16_t len = (uint16_t)*Len;
+    // Update circular buffer with incoming values
+    for(uint8_t i = 0; i < len; i++)
+    {
+        circular_buf_put(usb_cdc_if.rx.ctx, Buf[i]);
+
+        // Checking for '\r' is there for debugging purposes which is sent by putty and minicom.
+        // In production only '\n' is sent. Therefore, there is no check for '\r\n' case.
+        if (Buf[i]=='\n'||Buf[i]=='\r')
+            circular_buf_add_input(usb_cdc_if.rx.ctx);
+    }
+    memset(Buf, '\0', len); // clear the buf
+
+    return (USBD_OK);
+}
+
+
+/**
+  * @brief  usb_cdc_rx
+  *         Check/receive data from the USB cdc device.
+  * @param  buf: pointer to receiver buffer
+  * @param  Len: length of buffer
+  * @retval no of bytes received.
+  */
+void usb_cdc_rx(uint8_t* buf)
+{
+    *buf = 0;
+    if (usb_cdc_if.rx.ctx)
+        circular_read_command(usb_cdc_if.rx.ctx, buf);
+}
+
+void usb_cdc_flush()
```

Comment:
Consider renaming to something like "usb_rx_flush" since it only concerns receiver side or use an input argument to allow clearing of transmit side - even though I'm not sure when it would be relevant to flush transmit buffer :)